regex/src/main/scala/regex/Regex.scala

```diff
@@ -15,13 +29,51 @@ package regex
   */
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = 
+  lang match
+    case Concat(lang1, lang2) => 
+      if lang1 == Epsilon then simplify(lang2)
+      else if lang2 == Epsilon then simplify(lang1)
+      else if (lang1  == Empty || lang2 == Empty) then Empty
+      else Concat(simplify(lang1), simplify(lang2))
+
+    case Union(lang1, lang2) => 
+      if lang1 == Empty then simplify(lang2)
+      else if lang2 == Empty then simplify(lang1)
+      else Union(simplify(lang1), simplify(lang2))
+
+    case Star(x) => 
+      x match
+        case Epsilon => Epsilon
+        case Empty => Empty
+        case y => Star(simplify(y))
+      
+    case otherwise => otherwise
```

Comment:
I didn't realize you could deal with non-matching inputs to a pattern-match with this syntax (instead of having to use a wildcard). That's pretty cool